lib/utils/createNodeSubgraphAndDataSource.ts

```diff
@@ -0,0 +1,61 @@
+import { DocumentNode } from 'graphql';
+import { LocalGraphQLDataSource } from '@apollo/gateway';
+import type { Subgraph } from '@apollo/federation-internals';
+import gql from 'graphql-tag';
+
+import { buildSubgraph, isObjectType } from '@apollo/federation-internals';
+
+import createNodeService from 'lib/service';
+import type { NodeServiceConfig } from 'lib/service';
+
+import { nodeServiceEndPoint } from './constants';
+
+const createNodeExternalType = (name: string): DocumentNode =>
+  gql`
+    extend type ${name} implements Node @key(fields: "id") {
+      id: ID! @external
+    }
+  `;
+
+const getTypesImplementingNode = (
+  subgraphs: readonly Subgraph[],
+): Set<string> =>
+  subgraphs.reduce<Set<string>>(
+    (names, subgraph) =>
+      subgraph.schema.types().reduce<Set<string>>((acc, type) => {
+        if (isObjectType(type) && type.interfaceImplementation('Node')) {
+          acc.add(type.name);
+        }
+        return acc;
+      }, names),
+    new Set(),
+  );
+
+const getNodeExternalTypeDefs = (
+  subgraphs: readonly Subgraph[],
+): DocumentNode[] =>
+  Array.from(getTypesImplementingNode(subgraphs), name =>
+    createNodeExternalType(name),
+  );
+
+const createNodeSubgraphAndDataSource = (
+  nodeServiceConfig: NodeServiceConfig,
+  servicesSubgraphs: readonly Subgraph[],
+): [Subgraph, LocalGraphQLDataSource] => {
+  const typeDefs = getNodeExternalTypeDefs(servicesSubgraphs);
+  const schema = createNodeService({
+    ...nodeServiceConfig,
+    typeDefs: nodeServiceConfig.typeDefs
+      ? nodeServiceConfig.typeDefs.concat(typeDefs)
+      : typeDefs,
+  });
+  const dataSource = new LocalGraphQLDataSource(schema);
+  const subgraph = buildSubgraph(
+    nodeServiceEndPoint.name,
+    nodeServiceEndPoint.url,
+    dataSource.sdl(),
+  );
```

Comment:
When updating the dataSource, a different url must be used, otherwise the gateway will use the previous dataSource (
[source](https://github.com/apollographql/federation/blob/5fad357cc60b1849ea6069965ea7945b01d5ff5a/gateway-js/src/index.ts#L732)), thus using the nodeService defined for the old supergraph.

Reply:
@DDDKnightmare changed to use a hash, let me know if that solves the issue. Otherwise we'll need to create our own `GraphQLDataSource` and make it dynamic instead of using `LocalGraphQLDataSource`